Adding Dashboard CRD + Controller (#1352)
DataDog/datadog-operator: docs/datadog_dashboard.md

# Datadog Dashboards
This feature is in beta.

## Overview
The DatadogDashboard Custom Resource Definition (CRD) allows users to create [dashboards][1] using the Operator and manage them as Kubernetes resources.

## Prerequisites

- Datadog Operator v1.9+
- **[Helm][2]** for deploying the Datadog Operator
- **[`kubectl` CLI][3]** for installing a `DatadogDashboard`

Tests were performed on Kubernetes versions >= `v1.30.0`

> ## Adding a DatadogDashboard
> 
> To deploy a `DatadogDashboard` with the Datadog Operator, use the [`datadog-operator` Helm chart][3].
> 
> 1. Install the [Datadog Operator][4]:
> 
>    First, add the Datadog Helm chart with
> 
>     ```shell
>     helm repo add datadog https://helm.datadoghq.com
>     ```
> 
>     1. Run the following command, substituting your [Datadog API and application keys][5]:
> 
>         ```shell
>         helm install my-datadog-operator datadog/datadog-operator --set apiKey=<DATADOG_API_KEY> --set appKey=<DATADOG_APP_KEY> --set datadogDashboard.enabled=true
>         ```
> 
>     1. Alternatively, update the [`values.yaml`][6] file of the Datadog Operator Helm chart to include your [Datadog API and application keys][5] and enable `DatadogDashboard`.
>        Then, run
> 
>         ```shell
>         helm install my-datadog-operator datadog/datadog-operator -f values.yaml
>         ```
> 

Reviewer: I'm not sure how this renders, but linear steps are clearer. Where possible recommend keeping instructions at the same level (double check my spacing):
````suggestion
## Add a DatadogDashboard

To deploy a `DatadogDashboard` with the Datadog Operator, use the [`datadog-operator` Helm chart][3]. You can install the [Datadog Operator][4], or update an existing Datadog Operator Helm chart.

### Install the Datadog Operator
1. Add the Datadog Helm chart.
     ```shell
     helm repo add datadog https://helm.datadoghq.com
     ```
2. Run the following command, substituting your [Datadog API and application keys][5]:
     ```shell
        helm install my-datadog-operator datadog/datadog-operator --set apiKey=<DATADOG_API_KEY> --set appKey=<DATADOG_APP_KEY> --set datadogDashboard.enabled=true
     ```
### Update the Datadog Operator Helm chart
1. Update the [`values.yaml`][6] file of the Datadog Operator Helm chart to include your [Datadog API and application keys][5].
2. Enable `DatadogDashboard`.
3. Run the following command:
      ```shell
      helm install my-datadog-operator datadog/datadog-operator -f values.yaml
      ```

````

Reviewer 2: thanks @estherk15 ! we're removing this file for now but i will incorporate your suggestions on release in a few weeks.

# Datadog Dashboards
This feature will be introduced in DatadogOperator v1.9.0 and is currently in beta.

## Overview

> The Datadog Dashboards CRD allows users to create dashboards using the operator and manage them as Kubernetes resources. 

Reviewer: What does CRD stand for? Or is this a commonly used acronym?

Reviewer 3: Custom Resource Definition - it's commonly used and well understood acronym for Kubernetes users.

> All configurable Dashboard fields are available in the CRD: `Description`, `LayoutType`, `NotifyList`, `ReflowType`, `Tags`, `TemplateVariables`, `TemplateVariablePresets`, `Title`, and `Widgets`. (https://docs.datadoghq.com/dashboards/)

Reviewer: Can you verify this renders as a hyperlink?

```suggestion
All configurable Dashboard fields are available in the CRD: `Description`, `LayoutType`, `NotifyList`, `ReflowType`, `Tags`, `TemplateVariables`, `TemplateVariablePresets`, `Title`, and `Widgets`. For more information, see the [Dashboard](https://docs.datadoghq.com/dashboards/) documentation.
```

There are two main ways to define the Spec of a Datadog Dashboard. The first way is by specifying each field in a yaml file like so:

> ```

Reviewer: ```suggestion
```yaml
```

apiVersion: datadoghq.com/v1alpha1
kind: DatadogDashboard
metadata:
  name: example-dashboard
  namespace: system 
spec:
  title: metricsTest
  layoutType: ordered

  widgets:
    - timeseries:
        title: "itsTime"
        titleAlign: center
        titleSize: "250"
        requests: 
          - queries:
              - metricQuery: 
                  query: avg:system.cpu.user{*}
                  name: metrics
                  dataSource: metrics
                  aggregator: avg
            responseFormat: timeseries
            style:
              lineType: dotted
              lineWidth: thin
            onRightYaxis: false
      id: 124
```
However, because adding widgets is a very big pull, only two widget types (`Timeseries` and `QueryValue`) are supported this way.

> The second way is by copying the JSON of a dashboard in app, escaping it and just putting that into the `Data` field. Since this string isn't as easy to configure, there

Reviewer: ```suggestion
The second way is by copying the JSON of a dashboard in app, escaping it, and putting that into the `Data` field. Since this string isn't as easy to configure, there
```

spec:
  title: test
  layoutType: ordered
  tags:
    - "team:tagtest"
  templateVariables:
    - availableValues: 
        - host1
        - host2
        - host3
      name: first
      prefix: bar-foo
  notifyList:
    - foobar@datadoghq.com
  data:  "{\"title\":\"title\",\"description\":null,\"widgets\":[{\"id\":1234,\"definition\":{\"title\":\"\",\"type\":\"query_value\",\"requests\":[{\"queries\":[{\"compute\":{\"aggregation\":\"count\"},\"data_source\":\"logs\",\"indexes\":[\"*\"],\"name\":\"query1\",\"search\":{\"query\":\"\"},\"storage\":\"hot\"}],\"response_format\":\"scalar\"}],\"autoscale\":true,\"precision\":2,\"timeseries_background\":{\"type\":\"area\"}}}],\"template_variables\":[{\"name\":\"first\",\"prefix\":\"bar-foo\",\"available_values\":[\"host1\",\"host2\",\"host3\"],\"default\":\"*\"}],\"layout_type\":\"ordered\",\"notify_list\":[],\"template_variable_presets\":[{\"name\":\"saved_view1\",\"template_variables\":[{\"name\":\"foo-bar\",\"values\":[\"foo\",\"bar\"]}]}],\"reflow_type\":\"auto\",\"tags\":[\"team:tagtest\"]}"
```
Here, the created dashboard's title would be "test" instead of "title" (which is what is in the `Data` field).

## Prerequisites

> * Operator v1.9.0+

Reviewer: Is Operator v1.9.0 released? How do customers have access to the beta?

Reviewer 3: Not yet, but goal is to release this change in 1.9.

* Tests were performed on Kubernetes versions v>=`1.30.0`

> ## Enabling DatadogDashboards

Reviewer: Is this intentional? If so should it be `datadogDashboard` instead?
```suggestion
## Enabling Datadog Dashboards
```

> DAP is disabled by default. To enable DAP using the [datadog-operator helm chart](https://github.com/DataDog/helm-charts/tree/main/charts/datadog-operator), set `datadogDashboard.enabled=true` in your `values.yaml` file or as a flag in the command line arguments `--set datadogDashboard.enabled=true`.

Reviewer: What does DAP stand for?

Reviewer 3: It should be Dashboards, this section is missing something else too so will update it.
